src/libs/API.js

```diff
@@ -190,6 +191,17 @@ function request(command, parameters, type = 'post') {
                         return;
                     }
 
+                    if (!credentials || !credentials.autoGeneratedLogin || !credentials.autoGeneratedPassword) {
+                        // If we are here with invalid credentials it means we are trying to make an API request
+                        // before we are signed in. In this case, we should not continue.
+                        console.debug('A request tried to renew an authToken without credentials.', {
+                            command,
+                            parameters,
+                        });
+
+                        throw new Error('A request tried to renew an authToken without credentials.');
+                    }
```

Comment:
I'm confused about why we checking the credentials in the `.then()` at this point we have:

- An expired session
- The command we called is not `CreateLogin` or `Authenticate` since those would return early above (they are not retry-able see `parameters.doNotRetry`).

Reply:
At first I tried to insert it before `networkPromise`. 
But as first step of Sign in we send `GetAccountStatus` to check the email. And it's sending with expired session, so this check wouldn't allow it to go through.
So I decided to put it in `then` section instead.

Reply:
Hmm but `GetAccountStatus` does not require an `authToken` 🤔 

https://github.com/Expensify/Expensify.cash/blob/9f48370459159c7599be6f23661de737319739a0/src/libs/API.js#L21-L32

Reply:
Yes, but it's still being sent with this `request` function. So if I add the check for credentials before creating the promise, it will not go through.

But I think I can use this `isAuthTokenRequired` function to check the command before the request.
How about this?

```javascript
function request(command, parameters, type = 'post') {
    return new Promise((resolve, reject) => {
        if (isAuthTokenRequired(command) && (!credentials || !credentials.autoGeneratedLogin || !credentials.autoGeneratedPassword)) {
            // If we are here with invalid credentials it means we are trying to make an API request
            // before we are signed in. In this case, we should not continue.
            console.debug('A request tried to renew an authToken without credentials.', {
                command,
                parameters,
            });

            throw new Error('A request tried to renew an authToken without credentials.');
        }

        const networkPromise = mockCommands[command]
            ? mockCommands[command]()
            : Network.post(command, parameters, type);
```

Reply:
In this case probably `(!credentials || !credentials.autoGeneratedLogin || !credentials.autoGeneratedPassword)` should be extracted to function, because the condition becomes too big

Reply:
You mean something like this?

```javascript
                if (response.jsonCode === 407) {
                    const unableToReauthenticate = (!credentials || !credentials.autoGeneratedLogin || !credentials.autoGeneratedPassword);

                    if (parameters.doNotRetry || unableToReauthenticate) {
                        resolve(response);
                        return;
                    }

                    handleExpiredAuthToken(command, parameters, type)
```

Reply:
yes, exactly